docs/man/man8/lsadmin.8
.ds ]W %
.ds ]L
.nh
.TH lsadmin 8 "LSF Version 4.2 - June 2001"
.br
.SH NAME
\fBlsadmin\fR - administrative tool for LSF
.SH SYNOPSIS
.BR
.PP
.br
\fBlsadmin\fR \fIsubcommand\fR 
.br
\fBlsadmin\fR [\fB-h\fR | \fB-V\fR]
.SH SUBCOMMAND LIST
.BR
.PP
.br
\fBckconfig\fR [\fB-v\fR]
.br
\fBreconfig\fR [\fB-f\fR] [\fB-v\fR]
.br
\fBlimstartup\fR [\fB-f\fR] [\fIhost_name\fR ... | \fBall\fR]
.br
\fBlimshutdown\fR [\fB-f\fR] [\fIhost_name\fR ... | \fBall\fR]
.br
\fBlimrestart\fR [\fB-v\fR] [\fB-f\fR] [\fIhost_name\fR ... | \fBall\fR]
.br
\fBlimlock\fR [\fB-l\fR \fItime_seconds\fR]
.br
\fBlimunlock
\fR.br
\fBlimdebug\fR [\fB-c\fR \fIclass_name\fR ...] [\fB-l\fR \fIdebug_level\fR] [\fB-f\fR\fI logfile_name\fR] [\fB-o\fR] 
[\fIhost_name\fR]
.br
\fBlimtime\fR [\fB-l\fR \fItiming_level\fR] [\fB-f\fR \fIlogfile_name\fR] [\fB-o\fR] [\fIhost_name\fR]
.br
\fBresstartup\fR [\fB-f\fR] [\fIhost_name\fR ... | \fBall\fR]
.br
\fBresshutdown\fR [\fB-f\fR] [\fIhost_name\fR ... | \fBall\fR]
.br
\fBresrestart\fR [\fB-f\fR] [\fIhost_name\fR ... | \fBall\fR]
.br
\fBreslogon\fR [\fIhost_name\fR ... | \fBall\fR] [\fB-c\fR \fIcpu_time\fR]
.br
\fBreslogoff\fR [\fIhost_name\fR ... | \fBall\fR]
.br
\fBresdebug\fR [\fB-c\fR \fIclass_name\fR ...] [\fB-l\fR \fIdebug_level\fR] [\fB-f\fR \fIlogfile_name\fR] [\fB-o\fR] 
[\fIhost_name\fR]
.br
\fBrestime\fR [\fB-l\fR \fItiming_level\fR] [\fB-f\fR \fIlogfile_name\fR] [\fB-o\fR] [\fIhost_name\fR]
.br
\fBhelp\fR [\fIsubcommand\fR ...]
.br
\fBquit
\fR.SH DESCRIPTION
.BR
.PP
.PP
This command can only be used by LSF administrators. 
.PP
lsadmin is a tool that executes privileged commands to control LIM 
and RES operations in an LSF cluster. 
.PP
If no subcommands are supplied for lsadmin, lsadmin prompts for 
subcommands from the standard input.

.IP
For subcommands for which multiple host names or host groups can 
be specified, do not enclose the multiple names in quotation marks. 


.SH OPTIONS
.BR
.PP
.TP 
\fIsubcommand
\fR
.IP
Executes the specified subcommand. See Usage section. 


.TP 
\fB-h\fR 

.IP
Prints command usage to stderr and exits.


.TP 
\fB-V\fR 	 

.IP
Prints LSF release version and exits. 


.SH USAGE
.BR
.PP
.TP 
\fBckconfig \fR[\fB-v\fR]

.IP
Checks LSF configuration files.


.IP
\fB-v \fR
.BR
.RS
.IP
Displays detailed messages about configuration file checking. 

.RE

.TP 
\fBreconfig \fR[\fB-f\fR] [\fB-v\fR]

.IP
Restarts LIMs on all hosts in the cluster. You should use reconfig after 
changing configuration files. The configuration files are checked before 
all LIMs in the cluster are restarted. If the configuration files are not 
correct, reconfiguration will not be initiated.


.IP
\fB-f \fR
.BR
.RS
.IP
Disables user interaction and forces LIM to restart on all hosts 
in the cluster if no fatal errors are found. This option is useful 
in batch mode.

.RE

.IP
\fB-v\fR
.BR
.RS
.IP
Displays detailed messages about configuration file checking.

.RE

.TP 
\fBlimstartup \fR[\fB-f\fR] [\fIhost_name \fR... |\fBall\fR] 

.IP
Starts up the LIM on the local host if no arguments are specified.

.IP
Starts up LIMs on the specified hosts or on all hosts in the cluster if the 
word all is the only argument provided. You will be asked for 
confirmation.


.IP
\fB-f\fR
.BR
.RS
.IP
Disables interaction and does not ask for confirmation for 
starting up LIMs. \fB
\fR
.RE

.TP 
\fBlimshutdown \fR[\fB-f\fR] [\fIhost_name \fR... | \fBall\fR]

.IP
Shuts down LIM on the local host if no arguments are supplied.

.IP
Shuts down LIMs on the specified hosts or on all hosts in the cluster if 
the word all is specified. You will be asked for confirmation.


.IP
\fB-f\fR
.BR
.RS
.IP
Disables interaction and does not ask for confirmation for 
shutting down LIMs. 

.RE

.TP 
\fBlimrestart\fR [\fB-v\fR] [\fB-f\fR] [\fIhost_name \fR... | \fBall\fR]

.IP
Restarts LIM on the local host if no arguments are supplied.

.IP
Restarts LIMs on the specified hosts or on all hosts in the cluster if the 
word all is specified. You will be asked for confirmation.

.IP
limrestart should be used with care. Do not make any modifications 
until all the LIMs have completed the startup process. If you execute 
limrestart \fIhost_name\fR\fI...\fR to restart some of the LIMs after changing 
the configuration files, but other LIMs are still running the old 
configuration, confusion will arise among these LIMs. To avoid this 
situation, use reconfig instead of limrestart.


.IP
\fB-v\fR 
.BR
.RS
.IP
Displays detailed messages about configuration file checking.

.RE

.IP
\fB-f \fR
.BR
.RS
.IP
Disables user interaction and forces LIM to restart if no fatal 
errors are found. This option is useful in batch mode. 
limrestart -f all is the same as reconfig -f.

.RE

.TP 
\fBlimlock\fR [\fB-l\fR \fItime_seconds\fR]

.IP
Locks LIM on the local host until it is explicitly unlocked if no time is 
specified.When a host is locked, LIM's load status becomes lockU. No 
job will be sent to a locked host by LSF.


.IP
\fB-l \fR\fItime_seconds\fR
.BR
.RS
.IP
The host is locked for the specified time in seconds. This is 
useful if a machine is running an exclusive job requiring all the 
available CPU time and/or memory. 

.RE

.TP 
\fBlimunlock
\fR
.IP
Unlocks LIM on the local host. 


.TP 
\fBresstartup \fR[\fB-f\fR]\fB \fR[\fIhost_name\fR ... | \fBall\fR]

.IP
Starts up RES on the local host if no arguments are specified.

.IP
Starts up RESs on the specified hosts or on all hosts in the cluster if the 
word all is specified. You will be asked for confirmation.

For root installation to work properly, lsadmin must be installed as a 
setuid to root program. 


.IP
\fB-f\fR
.BR
.RS
.IP
Disables interaction and does not ask for confirmation for 
starting up RESs.

.RE

.TP 
\fBresshutdown\fR [\fB-f\fR] [\fIhost_name \fR... | \fBall\fR]

.IP
Shuts down RES on the local host if no arguments are specified. 

.IP
Shuts down RESs on the specified hosts or on all hosts in the cluster if 
the word all is specified. You will be asked for confirmation.

.IP
If RES is running, it will keep running until all remote tasks exit. 


.IP
\fB-f\fR
.BR
.RS
.IP
Disables interaction and does not ask for confirmation for 
shutting down RESs. 

.RE

.TP 
\fBresrestart \fR[\fB-f\fR] [\fIhost_name \fR... | \fBall\fR] 

.IP
Restarts RES on the local host if no arguments are specified. 

.IP
Restarts RESs on the specified hosts or on all hosts in the cluster if the 
word all is specified. You will be asked for confirmation.

.IP
If RES is running, it will keep running until all remote tasks exit. While 
waiting for remote tasks to exit, another RES is restarted to serve the 
new queries. 


.IP
\fB-f\fR
.BR
.RS
.IP
Disables interaction and does not ask for confirmation for 
restarting RESs. 

.RE

.TP 
\fBreslogon\fR [\fIhost_name \fR... | \fBall\fR] [\fB-c\fR \fIcpu_time\fR]

.IP
Logs all tasks executed by RES on the local host if no arguments are 
specified.

.IP
Logs tasks executed by RESs on the specified hosts or on all hosts in 
the cluster if all is specified. 

.IP
RES will write the task's resource usage information into the log file 
lsf.acct.\fIhost_name\fR. The location of the log file is determined by 
LSF_RES_ACCTDIR defined in lsf.conf. If LSF_RES_ACCTDIR is not 
defined, or RES cannot access it, the log file will be created in /tmp 
instead. 


.IP
\fB-c\fR \fIcpu_time\fR
.BR
.RS
.IP
Logs only tasks that use more than the specified amount of CPU 
time. The amount of CPU time is specified by \fIcpu_time\fR in 
milliseconds. 

.RE

.TP 
\fBreslogoff\fR [\fIhost_name \fR... |\fB all\fR]

.IP
Turns off RES task logging on the local host if no arguments are 
specified.

.IP
Turns off RES task logging on the specified hosts or on all hosts in the 
cluster if all is specified. 


.TP 
\fBlimdebug\fR [\fB-c\fR \fB"\fR\fIclass_name \fR...\fB"\fR] 
.br
[\fB-l\fR \fIdebug_level\fR] [\fB-f\fR \fIlogfile_name\fR] 
.br
[\fB-o\fR] [\fB"\fR\fIhost_name \fR...\fB"\fR]

.IP
Sets the message log level for LIM to include additional information in 
log files. You must be root or the LSF administrator to use this 
command. 

.IP
If the command is used without any options, the following default 
values are used:

.IP
\fIclass_name\fR = 0 (no additional classes are logged)

.IP
\fIdebug_level\fR = 0 (LOG_DEBUG level in parameter LSF_LOG_MASK)

.IP
\fIlogfile_name\fR = current LSF system log file in the directory specified by 
LSF_LOGDIR in the format \fIdaemon_name.\fRlog\fI.host_name
\fR
.IP
\fIhost_name\fR= local host (host from which command was submitted)


.IP
\fB-c\fR \fB"\fR\fIclass_name \fR...\fB"\fR
.BR
.RS
.IP
Specify software classes for which debug messages are to be 
logged. If a list of classes is specified, they must be enclosed in 
quotation marks and separated by spaces.

.IP
Possible classes: 

.IP
LC_AUTH - Log authentication messages

.IP
LC_CHKPNT - log checkpointing messages

.IP
LC_COMM - Log communication messages

.IP
LC_EXEC - Log significant steps for job execution

.IP
LC_FILE - Log file transfer messages

.IP
LC_HANG - Mark where a program might hang

.IP
LC_PIM - Log PIM messages

.IP
LC_SCHED - Log JobScheduler messages

.IP
LC_SIGNAL - Log messages pertaining to signals

.IP
LC_TRACE - Log significant program walk steps

.IP
LC_XDR - Log everything transferred by XDR

.IP
Default: 0 (no additional classes are logged)

.IP
Note: Classes are also listed in lsf.h.

.RE

.IP
\fB-l\fR \fIdebug_level\fR
.BR
.RS
.IP
Specify level of detail in debug messages. The higher the 
number, the more detail that is logged. Higher levels include all 
lower levels.

.IP
Possible values:

.IP
0 - LOG_DEBUG level in parameter LSF_LOG_MASK in 
lsf.conf. 

.IP
1 - LOG_DEBUG1 level for extended logging. A higher level 
includes lower logging levels. For example, LOG_DEBUG3 
includes LOG_DEBUG2 LOG_DEBUG1, and LOG_DEBUG 
levels.

.IP
2 - LOG_DEBUG2 level for extended logging. A higher level 
includes lower logging levels. For example, LOG_DEBUG3 
includes LOG_DEBUG2 LOG_DEBUG1, and LOG_DEBUG 
levels.

.IP
3 - LOG_DEBUG3 level for extended logging. A higher level 
includes lower logging levels. For example, LOG_DEBUG3 
includes LOG_DEBUG2, LOG_DEBUG1, and LOG_DEBUG 
levels.

.IP
Default: 0 (LOG_DEBUG level in parameter LSF_LOG_MASK)

.RE

.IP
\fB-f\fR \fIlogfile_name\fR
.BR
.RS
.IP
Specify the name of the file into which debugging messages are 
to be logged. A file name with or without a full path may be 
specified. 

.IP
If a file name without a path is specified, the file will be saved 
in the directory indicated by the parameter LSF_LOGDIR in 
lsf.conf.

.IP
The name of the file that will be created will have the following 
format:

.IP
\fIlogfile_name\fR.\fIdaemon_name\fR.log.\fIhost_name
\fR
.IP
If the specified path is invalid, on UNIX, the log file is created 
in the /tmp directory. 

.IP
If LSF_LOGDIR is not defined, daemons log to the syslog 
facility.

.IP
Default: current LSF system log file in the directory specified by 
LSF_LOGDIR in the format \fIdaemon_name\fR.log.\fIhost_name\fR.

.RE

.IP
\fB-o\fR
.BR
.RS
.IP
Turns off temporary debug settings and reset them to the 
daemon starting state. The message log level is reset back to the 
value of LSF_LOG_MASK and classes are reset to the value of 
LSF_DEBUG_RES, LSF_DEBUG_LIM.

.IP
Log file is reset back to the default log file.

.RE

.IP
\fB"\fR\fIhost_name \fR...\fB"\fR
.BR
.RS
.IP
Sets debug settings on the specified host or hosts.

.IP
Default: local host (host from which command was submitted)

.RE

.TP 
\fBresdebug\fR [\fB-c\fR \fB"\fR\fIclass_name\fR\fB"\fR] [\fB-l\fR \fIdebug_level\fR] [\fB-f\fR \fIlogfile_name\fR] [\fB-o\fR] 
[\fB"\fR\fIhost_name \fR...\fB"\fR]

.IP
Sets the message log level for RES to include additional information in 
log files. You must be the LSF administrator to use this command, not 
root.

.IP
See description of limdebug for an explanation of options.


.TP 
\fBlimtime\fR [\fB-l\fR \fItiming_level\fR] [\fB-f\fR \fIlogfile_name\fR] [\fB-o\fR] [\fB"\fR\fIhost_name ...\fR\fB"\fR]

.IP
Sets timing level for LIM to include additional timing information in log 
files. You must be root or the LSF administrator to use this command. 

.IP
If the command is used without any options, the following default 
values are used:

.IP
\fItiming_level\fR = no timing information is recorded

.IP
\fIlogfile_name\fR = current LSF system log file in the directory specified by 
LSF_LOGDIR in the format \fIdaemon_name\fR.log.\fIhost_name
\fR
.IP
\fIhost_name \fR= local host (host from which command was submitted)


.IP
\fB-l\fR \fItiming_level\fR
.BR
.RS
.IP
Specifies detail of timing information that is included in log 
files. Timing messages indicate the execution time of functions 
in the software and are logged in milliseconds.

.IP
Valid values: 1 | 2 | 3 | 4 | 5

.IP
The higher the number, the more functions in the software that 
are timed and whose execution time is logged. The lower 
numbers include more common software functions. Higher 
levels include all lower levels.

.IP
Default: undefined (no timing information is logged)

.RE

.IP
\fB-f\fR \fIlogfile_name\fR
.BR
.RS
.IP
Specify the name of the file into which timing messages are to 
be logged. A file name with or without a full path may be 
specified. 

.IP
If a file name without a path is specified, the file will be saved 
in the directory indicated by the parameter LSF_LOGDIR in 
lsf.conf.

.IP
The name of the file that will be created will have the following 
format:

.IP
\fIlogfile_name\fR.\fIdaemon_name\fR.log.\fIhost_name
\fR
.IP
If the specified path is invalid, on UNIX, the log file is created 
in the /tmp directory. 

.IP
If LSF_LOGDIR is not defined, daemons log to the syslog 
facility.

.IP
\fBNote: \fRBoth timing and debug messages are logged in the same 
files.

.IP
Default: current LSF system log file in the directory specified by 
LSF_LOGDIR in the format \fIdaemon_name\fR.log.\fIhost_name\fR.

.RE

.IP
\fB-o\fR
.BR
.RS
.IP
Turns off temporary timing settings and resets them to the 
daemon starting state. The timing level is reset back to the value 
of the parameter for the corresponding daemon 
(LSF_TIME_LIM, LSF_TIME_RES).

.IP
Log file is reset back to the default log file.

.RE

.IP
\fB"\fR\fIhost_name ...\fR\fB"\fR 
.BR
.RS
.IP
Sets the timing level on the specified host or hosts.

.IP
Default: local host (host from which command was submitted)

.RE

.TP 
\fBrestime\fR [\fB-l\fR \fItiming_level\fR] [\fB-f\fR \fIlogfile_name\fR] [\fB-o\fR] [\fB"\fR\fIhost_name ...\fR\fB"\fR]

.IP
Sets timing level for RES to include additional timing information in log 
files. You must be the LSF administrator can use this command, not 
root.

.IP
See description of limtime for an explanation of options.


.TP 
\fBhelp\fR [\fIsubcommand \fR...] | \fB?\fR [\fIsubcommand \fR...]

.IP
Displays the syntax and functionality of the specified commands. The 
commands must be explicit to lsadmin.

.IP
From the command prompt, you may use help or ?.


.TP 
\fBquit\fR 

.IP
Exits the lsadmin session. 


.SH SEE ALSO
.BR
.PP
.PP
ls_limcontrol(3), ls_rescontrol(3), ls_readconfenv(3), 
ls_gethostinfo(3), ls_connect(3), ls_initrex(3), 
lsf.conf(5), lsf.acct(5), bmgroup(1), 
busers(1)lsreconfig(8), lslockhost(8), lsunlockhost(8)


src/lsf/lsadm/lsadmin.c
/* $Id: lsadmin.c,v 1.9 2007/08/15 22:18:54 tmizan Exp $
 * Copyright (C) 2007 Platform Computing Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */

#include "lsf/lsadm/lsadm.h"

static int doLsCmd (int argc, char *argv[]);
static void limDebugReq(struct debugReq *pdebug, int num, char *host[]);
static void resDebugReq(struct debugReq *pdebug, int num, char *host[]);
static int lsadminDebug (int argc, char *argv[], int opCode);
static void support1_2 (int argc, char *argv[]);
static int checkPerm();
static int exitDebug;
extern int startup( int argc, char **argv, int opCode);
extern char *optarg;

static char *cmdList[] = {
    "reconfig",
    "ckconfig",
    "limrestart",
    "limstartup",
    "limshutdown",
    "limlock",
    "limunlock",
    "resrestart",
    "resstartup",
    "resshutdown",
    "reslogon",
    "reslogoff",
    "limdebug",
    "limtime",
    "resdebug",
    "restime",
    "help",
    "?",
    "quit",
    NULL
};

static char *cmdSyntax[] =
{
    "[-v] [-f]",
    "[-v]",
    "[-v] [-f] [ host_name ... | all ]",
    "[-f] [ host_name ... | all ]",
    "[-f] [ host_name ... | all ]",
    "[-l time_seconds] ",
    "",
    "[-f] [ host_name ... | all ]",
    "[-f] [ host_name ... | all ]",
    "[-f] [ host_name ... | all ]",
    "[ host_name ... | all ] [-c cpu_time]",
    "[ host_name ... | all ]",
    "[-c class_name] [-l debug_level] [-f logfile_name] [-o] [host_name ...]",
    "[-l timing_level] [-f logfile_name] [-o] [host_name ...]",
    "[-c class_name] [-l debug_level] [-f logfile_name] [-o] [host_name ...]",
    "[-l timing_level] [-f logfile_name] [-o] [host_name ...]",
    "[ command ...]",
    "[ command ...]",
    "",
    NULL
};

static char *cmdInfo[] = { "Reconfigure LSF",
                           "Check configuration files",
                           "Restart LIM (Load Information Manager)",
                           "Startup LIM (Load Information Manager)",
                           "Shut down LIM (Load Information Manager)",
                           "Lock LIM on local host",
                           "Unlock LIM on local host",
                           "Restart RES (Remote Execution Server)",
                           "Startup RES (Remote Execution Server)",
                           "Shut down RES (Remote Execution Server)",
                           "Log tasks executed by RES",
                           "RES log off",
                           "Debug LIM",
                           "Debug LIM timing message",
                           "Debug RES",
                           "Debug RES timing message",
                           "Get help on commands",
                           "Get help on commands",
                           "Quit",
                           NULL
};

static int opCodeList[] = {
    LIM_CMD_REBOOT,
    -1,
    LIM_CMD_REBOOT,
    -1,
    LIM_CMD_SHUTDOWN,
    0,
    -1,
    RES_CMD_REBOOT,
    -1,
    RES_CMD_SHUTDOWN,
    RES_CMD_LOGON,
    RES_CMD_LOGOFF,
    LIM_DEBUG,
    LIM_TIMING,
    RES_DEBUG,
    RES_TIMING,
    -1,
    -1,
    -1
};

int
main (int argc, char **argv)
{
    int cc,  myIndex;
    char *prompt = "lsadmin>";
    char line[MAXLINELEN];

    if (ls_initdebug(argv[0]) < 0) {
        ls_perror("ls_initdebug");
        exit(-1);
    }

    support1_2(argc, argv);

    while ((cc = getopt(argc, argv, "Vh")) != EOF) {
        switch (cc) {
        case 'V':
            fprintf(stderr, "%s\n", LAVALITE_VERSION_STR);
            return 0;
        case 'h':
        default:
            cmdsUsage("lsadmin", cmdList, cmdInfo);
        }
    }
    if (argc > optind) {
        if ((myIndex=adminCmdIndex(argv[optind], cmdList)) == -1) {
            fprintf(stderr, "Invalid command <%s> \n", argv[optind]);

            cmdsUsage("lsadmin", cmdList,
                      cmdInfo);
        }
        optind++;
        exit (doLsCmd (argc, argv));
    }

    for (;;) {
        printf("%s", prompt);
        fflush(stdout);
        if (fgets(line, MAXLINELEN, stdin) == NULL) {
            printf("\n");
            exit(-1);
        }
        parseAndDo (line , doLsCmd);

    }
}

static int
doLsCmd (int argc, char *argv[])
{
    int cmdRet = 0, cc, myIndex;

    if ((myIndex=adminCmdIndex(argv[optind-1], cmdList)) == -1) {
        fprintf(stderr, "Invalid command <%s>. Try help\n",  argv[optind-1]);
        return -1;
    }
    switch (myIndex) {
    case LSADM_CKCONFIG :
        if ((argc == optind + 1) && strcmp(argv[optind], "-v") == 0)
            cc =  checkConf(1, 1);
        else if (argc == optind) {
            cc = checkConf(0, 1);
            if (cc == -1 || cc == -2)
                if (getConfirm("Do you want to see detailed messages? [y/n] "))
                    cc =  checkConf(1, 1);
        } else
            cmdRet = -2;
        if (cmdRet == 0 && cc != 0)
            cmdRet = -1;
        break;
    case LSADM_RECONFIG :
    case LSADM_LIMREBOOT :
    case LSADM_LIMSHUTDOWN :
        cmdRet = limCtrl (argc, argv, opCodeList[myIndex]);
        break;
    case LSADM_LIMSTARTUP :
        cmdRet = startup (argc, argv, myIndex);
        break;
    case LSADM_LIMLOCK :
        cmdRet = limLock (argc, argv);
        break;
    case LSADM_LIMUNLOCK :
        cmdRet = limUnlock (argc, argv);
        break;
    case LSADM_RESREBOOT :
    case LSADM_RESSHUTDOWN :
    case LSADM_RESLOGON :
    case LSADM_RESLOGOFF :
        cmdRet = resCtrl (argc, argv, opCodeList[myIndex]);
        break;
    case LSADM_LIMDEBUG:
    case LSADM_LIMTIME:
    case LSADM_RESDEBUG:
    case LSADM_RESTIME:
        cmdRet = lsadminDebug (argc, argv, opCodeList[myIndex]);
        break;
    case LSADM_RESSTARTUP :
        cmdRet = startup (argc, argv, myIndex);
        break;
    case LSADM_HELP :
    case LSADM_QES :
        cmdHelp (argc, argv, cmdList, cmdInfo, cmdSyntax);
        break;
    case LSADM_QUIT :
        exit(0);
    default :
        fprintf(stderr, "adminCmdIndex:");
        exit(-1);
    }
    if (cmdRet == -2)
        oneCmdUsage(myIndex, cmdList, cmdSyntax);

    fflush(stderr);
    return (cmdRet);

}

static void
support1_2 (int argc, char *argv[])
{
    int cc;

    if (strstr(argv[0], "lsreconfig") != NULL) {
        if (argc != 1) {
            fprintf(stderr, "Error fix me %s\n", LAVALITE_VERSION_STR);
            exit(-1);
        }
        strcpy(argv[0], "reconfig");
        optind = 1;
        exit (limCtrl (1, argv, LIM_CMD_REBOOT));
    }
    if (strstr(argv[0], "lsunlockhost") != NULL) {
        if (argc != 1) {
            fprintf(stderr, "Error fix me %s\n", LAVALITE_VERSION_STR);
            exit(-1);
        }
        optind = 1;
        exit (limUnlock (1, argv));
    }
    if (strstr(argv[0], "lslockhost") != NULL) {
        while ((cc = getopt(argc, argv, "hV")) != EOF) {
            switch (cc) {
            case 'V':
                fprintf(stderr, "%s\n", LAVALITE_VERSION_STR);
                exit(0);
            default:
                fprintf(stderr, "lsadm: lslockhost [-V] [-h] [duration]\n");
                exit(-1);
            }
        }
        if (argc > optind)
            argv[--optind] = "-l";

        if ((cc=limLock (argc, argv)) == -2)
            fprintf(stderr, "lsadm: lslockhost [-V] [-h] [duration]\n");

        exit(cc);
    }
}

static int
lsadminDebug (int argc, char **argv, int opCode)
{
    struct hostInfo *hostInfo;
    char  opt[10];
    char *hosts[20];
    char *localHost;
    char *word;
    int fileOption = FALSE;
    int i,j, num, c;
    int   numHosts = 0;
    struct debugReq  debug;

    exitDebug = 0;
    debug.opCode = opCode ;
    debug.logClass = 0;
    debug.level = 0;
    debug.hostName = NULL;
    debug.options = 0;

    if (opCode == LIM_DEBUG || opCode == RES_DEBUG)
        strcpy(opt, "oc:l:f:");
    else if (opCode == LIM_TIMING || opCode == RES_TIMING)
        strcpy (opt, "ol:f:");
    else
        return (-2);

    while ((c = getopt(argc, argv, opt )) != EOF) {

        switch (c) {

        case 'c':
            while (optarg != NULL && (word = getNextWord_(&optarg))) {
                if (strcmp(word, "LC_SCHED") == 0)
                    debug.logClass |= LC_SCHED;
                if (strcmp(word, "LC_EXEC") == 0)
                    debug.logClass |= LC_EXEC;
                if (strcmp(word, "LC_TRACE") == 0)
                    debug.logClass |= LC_TRACE;
                if (strcmp(word, "LC_COMM") == 0)
                    debug.logClass |= LC_COMM;
                if (strcmp(word, "LC_XDR") == 0)
                    debug.logClass |= LC_XDR;
                if (strcmp(word, "LC_CHKPNT") == 0)
                    debug.logClass |= LC_CHKPNT;
                if (strcmp(word, "LC_FILE") == 0)
                    debug.logClass |= LC_FILE;
                if (strcmp(word, "LC_AUTH") == 0)
                    debug.logClass |= LC_AUTH;
                if (strcmp(word, "LC_HANG") == 0)
                    debug.logClass |= LC_HANG;
                if (strcmp(word, "LC_SIGNAL") == 0)
                    debug.logClass |= LC_SIGNAL;
                if (strcmp(word, "LC_PIM") == 0)
                    debug.logClass |= LC_PIM;
            }
            if (debug.logClass == 0) {
                fprintf(stderr,"Command denied. Invalid class name\n");
                return (-1);
            }
            break;
        case 'l':
            for (i=0;i<strlen(optarg);i++) {
                if (!isdigit(optarg[i])) {
                    fprintf(stderr,"Command denied. Invalid level value\n");
                    return(-1);
                }
            }
            debug.level = atoi(optarg);
            if (opCode == LIM_DEBUG || opCode == RES_DEBUG) {
                if (debug.level < 0 || debug.level > 3) {
                    fprintf(stderr,"Command denied. Valid debug level value is [0-3]\n");
                    return(-1);
                }
            }
            else if (debug.level < 1 || debug.level > 5) {
                fprintf(stderr,"Command denied. Valid timing level value is [1-5]\n");
                return(-1);
            }
            break;

        case 'f':
            if (strstr(optarg,"/") && strstr(optarg,"\\")) {
                fprintf(stderr,"Command denied. Invalid file name\n");
                return(-1);
            }
            fileOption = TRUE;
            memset(debug.logFileName,0,sizeof(debug.logFileName));
            ls_strcat(debug.logFileName,sizeof(debug.logFileName),optarg);
            if (debug.logFileName[strlen(debug.logFileName)-1] == '/' ||
                debug.logFileName[strlen(debug.logFileName)-1] == '\\') {
                fprintf(stderr,"Command denied. File name is needed after the path\n");
                return(-1);
            }
            break;

        case 'o':
            debug.options = 1;
            break;

        default:
            return (-2);
        }
    }

    if (checkPerm()) {
        if (opCode == LIM_DEBUG || opCode == LIM_TIMING ) {
            fprintf(stderr,"Operation permission denied by LIM\n");
        }
        else {
            fprintf(stderr,"Operation permission denied by RES\n");
        }
        return(-1);
    }
    if (opCode == LIM_DEBUG || opCode == LIM_TIMING ) {
        if ( !fileOption ) {
            strcpy (debug.logFileName, "lim");
        }
        else {
            strcat(debug.logFileName,".lim");
        }
    }
    else {
        if ( !fileOption ) {
            strcpy (debug.logFileName, "res");
        }
        else {
            strcat(debug.logFileName,".res");
        }
    }
    if (optind >= argc) {
        if ((localHost = ls_getmyhostname())  == NULL) {
            ls_perror("ls_getmyhostname");
            return -1;
        }
        if (opCode == LIM_DEBUG || opCode == LIM_TIMING ) {
            limDebugReq (&debug, 1, &localHost);
        }
        else {
            resDebugReq (&debug, 1, &localHost);
        }
        return (exitDebug);
    }

    if (optind == argc-1 && strcmp (argv[optind], "all") == 0) {

        hostInfo = ls_gethostinfo("-:server", &numHosts, NULL, 0, LOCAL_ONLY);
        if (hostInfo == NULL) {
            ls_perror("ls_gethostinfo");
            return (-1);
        }
        for (i=0; i<numHosts; i += 20) {
            num = (numHosts - i <  20) ? numHosts - i : 20;
            for (j=0; j<num; j++)
                hosts[j] = hostInfo[i+j].hostName;

            if (opCode == LIM_DEBUG || opCode == LIM_TIMING )
                limDebugReq (&debug, num, hosts);
            else
                resDebugReq (&debug, num, hosts);
        }
        return (exitDebug);
    }

    for (; optind < argc; optind += 20) {
        num = (argc - optind < 20) ? argc - optind : 20;
        if (opCode == LIM_DEBUG || opCode == LIM_TIMING )
            limDebugReq (&debug, num, &(argv[optind]));
        else
            resDebugReq (&debug, num, &(argv[optind]));
    }
    return(exitDebug);
}

static int
checkPerm(void)
{
    static char fname[] = "checkPerm";
    struct clusterInfo *clusterInfo;
    char  *mycluster = NULL;
    int   nAdmins = -1;
    int i;
    char username[128];

    if (getpwnam2(username) < 0) {
        return -1;
    }

    {
        struct passwd *pw;
        if ((pw = getpwuid((uid_t)getuid())) != NULL) {
            if (pw->pw_uid == 0)
                return(0);
        }
    }

    if ((mycluster = ls_getclustername()) == NULL) {
        ls_syslog(LOG_ERR, I18N_FUNC_FAIL_MM, fname, "ls_getclustername");
        return (RESE_DENIED);
    }
    if ((clusterInfo = ls_clusterinfo(NULL, NULL, NULL, 0, 0)) == NULL) {
        ls_syslog(LOG_ERR, I18N_FUNC_FAIL_MM, fname, "ls_clusterinfo");
        return (RESE_DENIED);
    }

    nAdmins = (clusterInfo->nAdmins) ? clusterInfo->nAdmins : 1;
    if (clusterInfo->nAdmins == 0) {
        if (strcmp(clusterInfo->managerName,username) == 0) {
            return(0);
        }
    }
    for (i = 0; i < nAdmins; i++) {
        if (strcmp(clusterInfo->admins[i],username) == 0) {
            return (0);
        }
    }
    return(-1);
}

static void
limDebugReq(struct debugReq *pdebug, int numHost, char *hosts[])
{
    int i;

    for (i=0; i<numHost; i++) {
        if (hosts[i][0] != '\0') {
            if(oneLimDebug(pdebug, hosts[i]) == -1) {
                fprintf(stderr, "Host %s set LIM debug failed", hosts[i]);
                exitDebug = -1;
            }
            else {
                fprintf(stderr, "Host %s set LIM debug done\n", hosts[i]);
            }
        }
    }
}

static void
resDebugReq(struct debugReq *pdebug, int numHost, char *hosts[])
{
    int pid;
    int i;
    int status;

    if ((pid = fork()) < 0) {
        perror("fork");
        exitDebug = -1;
        return ;
    }
    if (pid == 0) {
        if (ls_initrex(numHost, 0) < numHost ) {
            ls_perror ("ls_initrex:" );
            exit (-1);
        }

        for (i=0; i<numHost; i++)
            if (ls_connect(hosts[i]) < 0) {
                fprintf(stderr, "Host %s set RES debug failed", hosts[i]);
                exitDebug = -1;
                hosts[i][0] = '\0';
            }
        for (i=0; i<numHost; i++) {
            if (hosts[i][0] != '\0') {
                if (oneResDebug (pdebug,hosts[i] ) == -1) {
                    fprintf(stderr, "Host %s set RES debug failed: ", hosts[i]);
                    exitDebug = -1;
                }
                else {
                    fprintf(stderr, "Host %s set RES debug done\n", hosts[i]);
                }
            }
        }
        exit(exitDebug);
    }

    if (waitpid(pid, &status, 0) <0 ) {
        perror("waitpid");
        exitDebug = -1;
        return ;
    }

    if (WIFEXITED(status) == 0) {
        fprintf(stderr, "Child process killed by signal.\n");
        exitDebug = -1;
        return;
    }
    else {
        if (WEXITSTATUS(status) == 0xff) {
            exitDebug = -1;
            return;
        }
    }
}
